File: apps/api/app/pipeline/timing.py
from __future__ import annotations


DEFAULT_BEATS_PER_BAR = 4
DEFAULT_QUANTIZATION_SUBDIVISION = 4


def bpm_to_beat_duration(bpm: int) -> float:
    if bpm <= 0:
        return 0.0
    return 60.0 / bpm


def seconds_to_beats(seconds: float, bpm: int) -> float:
    beat_duration = bpm_to_beat_duration(bpm)
    if beat_duration <= 0:
        return 0.0
    return seconds / beat_duration


def beats_to_seconds(beats: float, bpm: int) -> float:
    return beats * bpm_to_beat_duration(bpm)


def quantize_beat(beat_position: float, subdivision: int = DEFAULT_QUANTIZATION_SUBDIVISION) -> float:
    if subdivision <= 0:
        return round(beat_position, 6)
    return round(round(beat_position * subdivision) / subdivision, 6)


def quantize_seconds(seconds: float, bpm: int, subdivision: int = DEFAULT_QUANTIZATION_SUBDIVISION) -> float:
    quantized_beats = quantize_beat(seconds_to_beats(seconds, bpm), subdivision)
    return round(beats_to_seconds(quantized_beats, bpm), 3)


def absolute_beat_to_bar_beat(absolute_beat: float, beats_per_bar: int = DEFAULT_BEATS_PER_BAR) -> tuple[int, float]:
    safe_beats_per_bar = max(1, beats_per_bar)
    beat_index = int(absolute_beat)
    bar = (beat_index // safe_beats_per_bar) + 1
    beat = round((absolute_beat % safe_beats_per_bar) + 1, 2)
    return bar, beat

File: packages/music-engine/src/timing.ts
import type { NoteEvent } from "@ai-sheet-music-generator/shared-types";

export const DEFAULT_BEATS_PER_BAR = 4;
export const DEFAULT_QUANTIZATION_SUBDIVISION = 4;

export function bpmToBeatDuration(bpm: number): number {
  if (bpm <= 0) {
    return 0;
  }

  return 60 / bpm;
}

export function secondsToBeats(seconds: number, bpm: number): number {
  const beatDuration = bpmToBeatDuration(bpm);
  if (beatDuration <= 0) {
    return 0;
  }

  return seconds / beatDuration;
}

export function beatsToSeconds(beats: number, bpm: number): number {
  return beats * bpmToBeatDuration(bpm);
}

export function quantizeBeat(beatPosition: number, subdivision = DEFAULT_QUANTIZATION_SUBDIVISION): number {
  if (subdivision <= 0) {
    return Number(beatPosition.toFixed(6));
  }

  return Number((Math.round(beatPosition * subdivision) / subdivision).toFixed(6));
}

export function quantizeSeconds(seconds: number, bpm: number, subdivision = DEFAULT_QUANTIZATION_SUBDIVISION): number {
  return Number(beatsToSeconds(quantizeBeat(secondsToBeats(seconds, bpm), subdivision), bpm).toFixed(3));
}

export function absoluteBeatToBarBeat(
  absoluteBeat: number,
  beatsPerBar = DEFAULT_BEATS_PER_BAR
): { bar: number; beat: number } {
  const safeBeatsPerBar = Math.max(1, beatsPerBar);
  const beatIndex = Math.floor(absoluteBeat);

  return {
    bar: Math.floor(beatIndex / safeBeatsPerBar) + 1,
    beat: Number(((absoluteBeat % safeBeatsPerBar) + 1).toFixed(2))
  };
}

export function formatBeatPosition(note: Pick<NoteEvent, "bar" | "beat">): string {
  if (note.bar == null || note.beat == null) {
    return "bar n/a beat n/a";
  }

  return `bar ${note.bar} beat ${note.beat}`;
}

export function formatEventTiming(note: Pick<NoteEvent, "onsetSec" | "offsetSec" | "bar" | "beat">): string {
  const end = note.offsetSec ?? note.onsetSec;
  return `${note.onsetSec.toFixed(2)}s to ${end.toFixed(2)}s | ${formatBeatPosition(note)}`;
}
